cabal-install/Distribution/Client/ProjectOrchestration.hs

```diff
@@ -223,8 +223,13 @@ data ProjectBuildContext = ProjectBuildContext {
 runProjectPreBuildPhase
     :: Verbosity
     -> ProjectBaseContext
-    -> (ElaboratedInstallPlan -> IO ElaboratedInstallPlan)
-    -> IO ProjectBuildContext
+    -- ^ A function that selects a subset of the plan and returns it,
+    -- It can also return some arbitrary data, which can be useful in case
+    -- the caller needs it.
+    -> (ElaboratedInstallPlan -> IO (ElaboratedInstallPlan, a))
```

Comment:
I don't have a strong preference on "return directly" vs. "stuff it in the `ProjectBuildStatus`, but for polymorphic vs. monomorphic, the polymorphic version is nice for forward-compatibility, e.g. if future consumers want some other information out of the pre-build phase target selection process.

If we stick with "return directly", perhaps there should be, like, a `runProjectPreBuildPhase_ f = fst <$> runProjectPreBuildPhase (\x -> (f x, ())` or something like that which has the old type. Then existing uses would be a bit less noisy.